### PR TITLE
feature: added capability to customize textinput inside Select. Also …

### PR DESCRIPTION
**Pull request by alansouzati**
…added better customization for checkbox.

#### What does this PR do?

Allows you to have a custom TextInput inside Select

#### Where should the reviewer start?

select.stories.js (custom search example)

#### What testing has been done on this PR?

manual

#### Do the grommet docs need to be updated?

We need to update the site CI, but the documentation inside grommet core has been updated

#### Should this PR be mentioned in the release notes?

Yes

#### Is this change backwards compatible or is it a breaking change?

Backwards compatible.

### CODE REVIEW

**Comment by alansouzati** [comment]
1) Fixed bug in accordion stories
2) Adding both `min-width: 0` and `min-height: 0` always. This helps to avoid having to set a given direction for Box just to fix flexbugs
3) Fixed focus management for CheckBox to align with our other inputs
4) Allowing a different hover style for CheckBox
5) Added `closeOnChange` option to Select
6) Fixed SelectContainer to use forwardRefs
7) Attaching mouseActive to the window object to avoid problems with portals
8) Added a default maxHeight to the Select drop
9) Improved select to have a smooth scroll when using the up and down keys. Now, we only change the scroll position if the target element is out of the scroll position.

File: src/js/components/hocs.js [hunk 1]
```diff
@@ -51,19 +51,19 @@ export const withFocus = (WrappedComponent) => {
     }
 
     handleActiveMouse = () => {
-      this.mouseActive = true;
+      window.mouseActive = true;
```
Comment (ericsoderberghp) [inline src/js/components/hocs.js]: Why does this need to be window?

File: src/js/components/Accordion/accordion.stories.js [hunk 2]
```diff
@@ -283,7 +283,7 @@ const renderPanelHeader = (title, active) => (
 
 class CustomHeaderAccordion extends Component {
   state = {
-    activeIndex: 0,
+    activeIndex: [0],
```
[the remark this answers is not in the document]
Comment (alansouzati) [reply src/js/components/Accordion/accordion.stories.js]: it still does. but as per you feedback in a previous pr, we are always sending the array on change. remember? 😄

File: src/js/components/Box/StyledBox.js [hunk 3]
```diff
@@ -59,8 +59,8 @@ const basisStyle = css`
 // we assume we are in the context of a Box going the other direction
 // TODO: revisit this
 const directionStyle = css`
-  ${props => props.directionProp === 'row' && 'min-height: 0;'}
-  ${props => props.directionProp === 'column' && 'min-width: 0;'}
+  min-width: 0;
```
[the remark this answers is not in the document]
Comment (alansouzati) [reply src/js/components/Box/StyledBox.js]: `Adding both min-width: 0 and min-height: 0 always. This helps to avoid having to set a given direction for Box just to fix flexbugs
Fixed focus management for CheckBox to align with our other inputs`

[the remark this answers is not in the document]
Comment (alansouzati) [reply src/js/components/Box/StyledBox.js]: without it I'm facing overflow issues when using `<Text truncate={true} />`

File: src/js/components/CheckBox/CheckBox.js [hunk 4]
```diff
@@ -27,15 +37,21 @@ class CheckBox extends Component {
...
       </StyledCheckBoxToggle>
     ) : (
-      <StyledCheckBoxBox theme={theme}>
-        <svg viewBox='0 0 24 24' preserveAspectRatio='xMidYMid meet'>
-          <path fill='none' d='M6,11.3 L10.3,16 L18,6.2' />
-        </svg>
+      <StyledCheckBoxBox focus={focus} theme={theme}>
+        {
+          Icon ? <Icon /> : (
+            <svg viewBox='0 0 24 24' preserveAspectRatio='xMidYMid meet'>
```
[the remark this answers is not in the document]
Comment (alansouzati) [reply src/js/components/CheckBox/CheckBox.js]: we currently don't export this icon anywhere? can we change it to use grommet-icons icon ?

File: src/js/components/hocs.js (hunk 1 above)
Comment (alansouzati) [reply src/js/components/hocs.js]: `Attaching mouseActive to the window object to avoid problems with portals`

Comment (alansouzati) [reply src/js/components/hocs.js]: mouseActive is getting set to false when opening a portal because `this.mouseActive` in a new portal is always false.

File: src/js/components/DropButton/__tests__/__snapshots__/DropButton-test.js.snap [hunk 5]
```diff
@@ -47,13 +47,6 @@ exports[`DropButton close by clicking outside 3`] = `
     -webkit-transition: 0.1s ease-in-out;
     transition: 0.1s ease-in-out;
   }
-}
-
-@media only screen and (min-width:700px) {
```
[the remark this answers is not in the document]
Comment (alansouzati) [reply src/js/components/DropButton/__tests__/__snapshots__/DropButton-test.js.snap]: not sure where it moved to. but we did not remove it.

https://github.com/grommet/grommet/blob/571a977eff5350d21716aed27ff4bb8b55453ded/src/js/components/DropButton/__tests__/__snapshots__/DropButton-test.js.snap#L218

File: src/js/components/Box/StyledBox.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (ericsoderberghp) [reply src/js/components/Box/StyledBox.js]: I seem to recall some situations where this isn't the right behavior. But, I'm having trouble remembering the details :(
This is the one change I am most nervous about causing an unexpected impact.

File: src/js/components/CheckBox/CheckBox.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (ericsoderberghp) [reply src/js/components/CheckBox/CheckBox.js]: We should add a story for this, since this would benefit from some visual verification.

File: src/js/components/Select/SelectContainer.js [hunk 6]
```diff
@@ -18,37 +25,35 @@ class SelectContainer extends Component {
...
-
   onInput = (event) => {
     this.setState(
       { search: event.target.value },
       () => this.onSearch(this.state.search)
     );
   }
 
-  onSearch = debounce(search => this.props.onSearch(search), 150)
+  onSearch = debounce(search => this.props.onSearch(search), 300)
```
Comment (ericsoderberghp) [inline src/js/components/Select/SelectContainer.js]: Can you include a comment on why we chose `300`?

File: src/js/components/Select/SelectContainer.js [hunk 7]
```diff
@@ -131,31 +153,35 @@ class SelectContainer extends Component {
...
                 placeholder={searchPlaceholder}
                 onInput={this.onInput}
               />
             </Box>
-          ) : undefined}
-
+          )}
           <Box
-            flex={false}
+            flex={true}
```
Comment (ericsoderberghp) [inline src/js/components/Select/SelectContainer.js]: This seems like a notable change. Why was this changed?

File: src/js/components/Select/SelectContainer.js [hunk 8]
```diff
@@ -131,31 +153,35 @@ class SelectContainer extends Component {
...
             role='menubar'
             tabIndex='-1'
-            ref={(ref) => { this.selectRef = ref; }}
+            ref={this.selectRef}
+            style={{
```
Comment (ericsoderberghp) [inline src/js/components/Select/SelectContainer.js]: Should these be in a styled component instead of inline styles?

File: src/js/components/Box/StyledBox.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (alansouzati) [reply src/js/components/Box/StyledBox.js]: I see. hum I cannot think of a way to fix that without creating a wrapper box, just to avoid this problem. I would much rather know for sure that we don't want to always use both.

File: src/js/components/Select/SelectContainer.js (hunk 7 above)
Comment (alansouzati) [reply src/js/components/Select/SelectContainer.js]: I'm not sure why we added flex false before. but I'm certain that it was to avoid elements collapsing.
But it makes sense for the search input to be fixed and the list of options to scroll. the list of options should take the remaining space in the drop height.

File: src/js/components/Select/SelectContainer.js (hunk 8 above)
Comment (alansouzati) [reply src/js/components/Select/SelectContainer.js]: agreed. let me change that.

File: src/js/components/Box/StyledBox.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (ericsoderberghp) [reply src/js/components/Box/StyledBox.js]: I've tested with with OneSphere and grommet-site. I'm not seeing any issues. So, I'm OK with this. Something might rear it's head later, but I don't want to hold this up for that.
